lib/pac55xx/can.c

```diff
@@ -0,0 +1,460 @@
+	word = ((~id_mask << 3) & CAN_ACR_SINGLE_EXT_ID);
+	CAN_AMR(canport) = __builtin_bswap32(word);
+	/* 1: single filter */
+	CAN_ISR_SR_CMR_MR_SET(canport, CAN_MR_AFM);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Enable IRQ
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] irq Unsigned int8. IRQ bit(s).
+*/
+void can_enable_irq(uint32_t canport, uint8_t irq) {
+	/* set to 1 (not masked) to enable */
+	CAN_BTR1_BTR0_RMC_IMR(canport) |= (uint32_t)irq;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Disable IRQ
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] irq Unsigned int8. IRQ bit(s).
+*/
+void can_disable_irq(uint32_t canport, uint8_t irq) {
+	/* set to 0 (masked) to disable */
+	CAN_BTR1_BTR0_RMC_IMR(canport) &= ~(uint32_t)irq;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Transmit Standard Frame
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] id Unsigned int32. Message ID bits 10:0 used.
+@param[in] rtr bool. Remote Request bit value.
+@param[in] length Unsigned int8. Message payload length.
+@param[in] data Unsigned int8[]. Message payload data.
+@returns true if able to transmit, false otherwise.
+*/
+bool can_transmit_std(uint32_t canport, uint32_t id, bool rtr, uint8_t length,
+			const uint8_t *data) {
+	/* if TBS is 0, then not ready to transmit */
+	if ((CAN_ISR_SR_CMR_MR(canport) & CAN_SR_TBS) == 0) {
+		return false;
+	}
+	uint32_t word = (length & CAN_BITS_3_0)
+		| (rtr ? BIT6 : 0)            /* DLC/RTR/FF ==> 7:0 */
+		| ((id & CAN_BITS_10_3) << 5) /* ID 10:3 ==> 15:8 */
+		| ((id & CAN_BITS_2_0) << 21) /* ID 2:0 ==> 23:21 */
+		| (((length > 0) ? data[0] : 0) << 24);
+	CAN_TXBUF(canport) = word;
+
+	if (length > 1) {
+		word = (data[1] << 0) | (data[2] << 8)
+			| (data[3] << 16) | (data[4] << 24);
+		CAN_TXBUF(canport) = word;
+	}
+
+	if (length > 5) {
+		word = (data[5] << 0) | (data[6] << 8) | (data[7] << 16);
+		CAN_TXBUF(canport) = word;
+	}
+
+	/* Request transmit */
+	CAN_ISR_SR_CMR_MR_SET(canport, CAN_CMR_TR);
+	return true;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief CAN Transmit Extended Frame
+@param[in] canport Unsigned int32. CAN block register base address.
+@param[in] id Unsigned int32. Message ID bits 28:0 used.
+@param[in] rtr bool. Remote Request bit value.
+@param[in] length Unsigned int8. Message payload length, 0-8.
+@param[in] data Unsigned int8[]. Message payload data.
+@returns true if able to transmit, false otherwise.
+*/
+bool can_transmit_ext(uint32_t canport, uint32_t id, bool rtr, uint8_t length,
+			const uint8_t *data) {
+	/* if TBS is 0, then not ready to transmit */
+	if ((CAN_ISR_SR_CMR_MR(canport) & CAN_SR_TBS) == 0) {
+		return false;
+	}
+	uint32_t word = (length & CAN_BITS_3_0)
```

Comment:
this too, BITS_3_0 isn't a helpful definition.  you've just got different magic.  this is "CAN_DLC_MASK"